File: phoenix_IV_functions/kalman_update.cpp
#include "phoenix_IV_functions.h"

void kalman_update(double *state, double p_cov[3][3], double *measurement, double dt, double *adjusted_state, double adjusted_p_cov[3][3]){

    double A[3][3] = {{1, dt, dt*dt / 2}, {0, 1, dt}, {0, 0, 1}};     //state transition matrix for position
    double Q[3][3] = {{3, 0, 0}, {0, 2, 0}, {0, 0, 1}};    //model noise covariance
    double H[2][3] = {{1, 0, 0}, {0, 0, 1}};     //measurement jacobian (to transforms the system state into the measurement)
    double R[2][2] = {{100, 0}, {0, 10}};  //measurement noise covariance
    double kalman_gain[3][2];


    predict_state(state, A, state);
    predict_p_cov(p_cov, A, Q, p_cov);
    update_gain(p_cov, H, R, kalman_gain);
    printf("%f %f \n %f%f \n %f %f \n", kalman_gain[0][0], kalman_gain[0][1], kalman_gain[1][0], kalman_gain[1][1], kalman_gain[2][0], kalman_gain[2][1]);
    adjust_state(state, measurement, H, kalman_gain, adjusted_state);
    adjust_p_cov(p_cov, kalman_gain, H, adjusted_p_cov);
}

File: rowans_kalman_filter_testing/MATLAB/kalman_update.m
function [state, p_cov] = kalman_update(state, p_cov, measurement, dt, drag_term)

    %https://www.bzarg.com/p/how-a-kalman-filter-works-in-pictures/
    
    global R Q MU;
    
    A = [1 dt ; 0 1];
    B = [0.5 * dt^2 ; dt];
    

    
    H = [1 0 ; 0 1];

    predicted_state = predict_state(state, A, B);
    
    predicted_p_cov = predict_p_cov(p_cov, A);
    
    kalman_gain = update_gain(predicted_p_cov, H, R);
    
    state = adjust_state(predicted_state, measurement, H, kalman_gain);
    
    p_cov = adjust_p_cov(predicted_p_cov, kalman_gain, H);
    


    
    function res = predict_state(state, A, B)
        drag_force = 0.5 * 1.5 * state(2)^2 * 0.2 * 1.225;
        drag_force = drag_force * (state(2) / abs(state(2)));
        drag = drag_force / 22;
        if drag_term
            res = A*state + B*(MU-drag);
        else
            res = A*state + B*MU;
        end
    end

    function res = predict_p_cov(p_cov, A)
        res = A * p_cov * A' + Q;
    end

    function res = update_gain(predicted_p_cov, H, R)
        kgain = (predicted_p_cov*H') / (H*predicted_p_cov*H' + R);
        res = kgain;
    end

    function res = adjust_state(state, measurement, H, kalman_gain)
        res = state + kalman_gain*(measurement - H*state);
    end

    function res = adjust_p_cov(predicted_p_cov, kalman_gain, H)
        res = eye(2,2) - kalman_gain*H*predicted_p_cov;
    end
end
